<a href="https://colab.research.google.com/github/22f1000003/Machine-Learnig-Practice/blob/week-1-machine-learning-prctice/satyam_Introduction_to__pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This is my**first experience** with the colab
the following is the list
* Item 1
* Item 2
   * subitem 1
   * subitem2

1.   List item
2.   List item
$$ v = w0 + w1×1$$  


In [ ]:

import pandas as pd
pd.__version__



'2.2.2'

The primary data structure in pandas are implemented as two classes:
* DataFrame. which you can imagin as a relational data table with rows and named columns.
* Series. which is a single column. Each row can be labled via an index.
* A DataFrame contains one or more Series and a name for each series.
The data frame is a commonly used abstration for data manipulation. similar implimentations exist in Sparkand R.

In [ ]:
from sklearn.datasets import load_diabetes
diabetes = load_diabetes( as_frame=True)
print(type(diabetes['data']))
print(diabetes.data_filename)


<class 'pandas.core.frame.DataFrame'>
diabetes_data_raw.csv.gz


In [ ]:
?load_diabetes

In [ ]:
df = diabetes['data']

In [ ]:
print(diabetes.keys())

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])


In [ ]:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

*Creating pandas dataframe from Series*

In [ ]:
cities = pd.Series(['delhi','Bengalore', 'Chennai','kanpur'])
population = pd.Series([17000000, 12000000, 6000000,])



**A DataFrame in pandas is:

👉 like a dictionary of columns (Series), where each column is like an array**

In [ ]:
city_info_df = pd.DataFrame({'City':cities , 'Population':population})


In [ ]:
type(city_info_df)

pandas.core.frame.DataFrame

In [ ]:
city_info_df

,City,Population
0,delhi,17000000.0
1,Bengalore,12000000.0
2,Chennai,6000000.0
3,kanpur,NaN


In [ ]:
city_info_df.head()

,City,Population
0,delhi,17000000.0
1,Bengalore,12000000.0
2,Chennai,6000000.0
3,kanpur,NaN


* psutil (process and system utilities)
* is a Python library for retrieving information on running processes and system utilization (CPU, memory, disks, network, sensors, etc.) in a cross-platform way (Linux, Windows, macOS, BSD, Solaris).

In [ ]:
import psutil
print(f"Cloud RAM used: {psutil.virtual_memory().percent}%")

Cloud RAM used: 9.5%


**Eploring Data in DataFrame**





Quikly examine few entries in the DataFrame. Say first five 5 and last 5.

In [ ]:
df.head()

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
0,0.038076,0.050680,0.061696,0.021872,-0.044223,-0.034821,-0.043401,-0.002592,0.019907,-0.017646
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204
2,0.085299,0.050680,0.044451,-0.005670,-0.045599,-0.034194,-0.032356,-0.002592,0.002861,-0.025930
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362
4,0.005383,-0.044642,-0.036385,0.021872,0.003935,0.015596,0.008142,-0.002592,-0.031988,-0.046641


In [ ]:
df.tail()

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
437,0.041708,0.050680,0.019662,0.059744,-0.005697,-0.002566,-0.028674,-0.002592,0.031193,0.007207
438,-0.005515,0.050680,-0.015906,-0.067642,0.049341,0.079165,-0.028674,0.034309,-0.018114,0.044485
439,0.041708,0.050680,-0.015906,0.017293,-0.037344,-0.013840,-0.024993,-0.011080,-0.046883,0.015491
440,-0.045472,-0.044642,0.039062,0.001215,0.016318,0.015283,-0.028674,0.026560,0.044529,-0.025930
441,-0.045472,-0.044642,-0.073030,-0.081413,0.083740,0.027809,0.173816,-0.039493,-0.004222,0.003064


In [ ]:
? df.head

In [ ]:
df.head(n=5)

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
0,0.038076,0.050680,0.061696,0.021872,-0.044223,-0.034821,-0.043401,-0.002592,0.019907,-0.017646
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204
2,0.085299,0.050680,0.044451,-0.005670,-0.045599,-0.034194,-0.032356,-0.002592,0.002861,-0.025930
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362
4,0.005383,-0.044642,-0.036385,0.021872,0.003935,0.015596,0.008142,-0.002592,-0.031988,-0.046641


In [ ]:
df.tail(n=3)

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
439,0.041708,0.050680,-0.015906,0.017293,-0.037344,-0.013840,-0.024993,-0.011080,-0.046883,0.015491
440,-0.045472,-0.044642,0.039062,0.001215,0.016318,0.015283,-0.028674,0.026560,0.044529,-0.025930
441,-0.045472,-0.044642,-0.073030,-0.081413,0.083740,0.027809,0.173816,-0.039493,-0.004222,0.003064


Find out number of rows and column in the dataframe

In [ ]:
df.shape

(442, 10)

* in this diabetea data i have 442 rows and 10  columns
* how to see the names of the columns

In [ ]:
df.columns

Index(['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6'], dtype='object')

In [ ]:
list(df.columns)

['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']

In [ ]:
?load_diabetes


In [ ]:
?DESCR

Object `DESCR` not found.


In [ ]:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

**Lecture-3**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442 entries, 0 to 441
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     442 non-null    float64
 1   sex     442 non-null    float64
 2   bmi     442 non-null    float64
 3   bp      442 non-null    float64
 4   s1      442 non-null    float64
 5   s2      442 non-null    float64
 6   s3      442 non-null    float64
 7   s4      442 non-null    float64
 8   s5      442 non-null    float64
 9   s6      442 non-null    float64
dtypes: float64(10)
memory usage: 34.7 KB


let's get summary statistics on the dataframe

In [ ]:
df.describe()

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
count,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02
mean,-2.511817e-19,1.230790e-17,-2.245564e-16,-4.797570e-17,-1.381499e-17,3.918434e-17,-5.777179e-18,-9.042540e-18,9.293722e-17,1.130318e-17
std,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02
min,-1.072256e-01,-4.464164e-02,-9.027530e-02,-1.123988e-01,-1.267807e-01,-1.156131e-01,-1.023071e-01,-7.639450e-02,-1.260971e-01,-1.377672e-01
25%,-3.729927e-02,-4.464164e-02,-3.422907e-02,-3.665608e-02,-3.424784e-02,-3.035840e-02,-3.511716e-02,-3.949338e-02,-3.324559e-02,-3.317903e-02
50%,5.383060e-03,-4.464164e-02,-7.283766e-03,-5.670422e-03,-4.320866e-03,-3.819065e-03,-6.584468e-03,-2.592262e-03,-1.947171e-03,-1.077698e-03
75%,3.807591e-02,5.068012e-02,3.124802e-02,3.564379e-02,2.835801e-02,2.984439e-02,2.931150e-02,3.430886e-02,3.243232e-02,2.791705e-02
max,1.107267e-01,5.068012e-02,1.705552e-01,1.320436e-01,1.539137e-01,1.987880e-01,1.811791e-01,1.852344e-01,1.335973e-01,1.356118e-01


In [ ]:
df.describe?

In [ ]:
df.describe(percentiles= [0.2,0.3,0.5,0.6,0.8])

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
count,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02
mean,-2.511817e-19,1.230790e-17,-2.245564e-16,-4.797570e-17,-1.381499e-17,3.918434e-17,-5.777179e-18,-9.042540e-18,9.293722e-17,1.130318e-17
std,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02
min,-1.072256e-01,-4.464164e-02,-9.027530e-02,-1.123988e-01,-1.267807e-01,-1.156131e-01,-1.023071e-01,-7.639450e-02,-1.260971e-01,-1.377672e-01
20%,-4.547248e-02,-4.464164e-02,-4.048038e-02,-4.009893e-02,-3.871969e-02,-3.695017e-02,-3.971921e-02,-3.949338e-02,-4.117617e-02,-3.835666e-02
30%,-2.730979e-02,-4.464164e-02,-2.991782e-02,-2.873752e-02,-2.633611e-02,-2.480001e-02,-2.867429e-02,-3.949338e-02,-2.796492e-02,-2.468771e-02
50%,5.383060e-03,-4.464164e-02,-7.283766e-03,-5.670422e-03,-4.320866e-03,-3.819065e-03,-6.584468e-03,-2.592262e-03,-1.947171e-03,-1.077698e-03
60%,1.628068e-02,5.068012e-02,5.218854e-03,8.100982e-03,8.062710e-03,8.706873e-03,8.142084e-03,-2.592262e-03,1.255119e-02,7.206516e-03
80%,4.170844e-02,5.068012e-02,4.229559e-02,4.941519e-02,3.943444e-02,3.952068e-02,3.759519e-02,3.430886e-02,3.885335e-02,3.620126e-02
max,1.107267e-01,5.068012e-02,1.705552e-01,1.320436e-01,1.539137e-01,1.987880e-01,1.811791e-01,1.852344e-01,1.335973e-01,1.356118e-01


here we transpose it by using T attribute so it is become more readeable

In [ ]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,442.0,-2.511817e-19,0.047619,-0.107226,-0.037299,0.005383,0.038076,0.110727
sex,442.0,1.230790e-17,0.047619,-0.044642,-0.044642,-0.044642,0.050680,0.050680
bmi,442.0,-2.245564e-16,0.047619,-0.090275,-0.034229,-0.007284,0.031248,0.170555
bp,442.0,-4.797570e-17,0.047619,-0.112399,-0.036656,-0.005670,0.035644,0.132044
s1,442.0,-1.381499e-17,0.047619,-0.126781,-0.034248,-0.004321,0.028358,0.153914
s2,442.0,3.918434e-17,0.047619,-0.115613,-0.030358,-0.003819,0.029844,0.198788
s3,442.0,-5.777179e-18,0.047619,-0.102307,-0.035117,-0.006584,0.029312,0.181179
s4,442.0,-9.042540e-18,0.047619,-0.076395,-0.039493,-0.002592,0.034309,0.185234
s5,442.0,9.293722e-17,0.047619,-0.126097,-0.033246,-0.001947,0.032432,0.133597
s6,442.0,1.130318e-17,0.047619,-0.137767,-0.033179,-0.001078,0.027917,0.135612


**Selection of particular column s and its attribute and i can say here geting series**

In [ ]:
from sklearn.datasets import load_diabetes
diabetes = load_diabetes(as_frame=True)
df = diabetes['data']

In [ ]:
df.columns

Index(['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6'], dtype='object')

In [ ]:
df['age']

,age
0,0.038076
1,-0.001882
2,0.085299
3,-0.089063
4,0.005383
...,...
437,0.041708
438,-0.005515
439,0.041708
440,-0.045472


In [ ]:
type?

In [ ]:
type(df['age'])

pandas.core.series.Series

how to find particular value in particular series

In [ ]:
df['age'][0]

np.float64(0.038075906433423026)

In [ ]:
df['age'][:5]

,age
0,0.038076
1,-0.001882
2,0.085299
3,-0.089063
4,0.005383


In [ ]:
df['age'][-5:]

,age
437,0.041708
438,-0.005515
439,0.041708
440,-0.045472
441,-0.045472


In [ ]:
df['age'][100:200]

,age
100,0.016281
101,0.016281
102,-0.092695
103,0.059871
104,-0.027310
...,...
195,0.027178
196,-0.023677
197,0.048974
198,-0.052738


In [ ]:
df[['age','sex']]

,age,sex
0,0.038076,0.050680
1,-0.001882,-0.044642
2,0.085299,0.050680
3,-0.089063,-0.044642
4,0.005383,-0.044642
...,...,...
437,0.041708,0.050680
438,-0.005515,0.050680
439,0.041708,0.050680
440,-0.045472,-0.044642


In [ ]:
df[['age','sex']][:5]

,age,sex
0,0.038076,0.050680
1,-0.001882,-0.044642
2,0.085299,0.050680
3,-0.089063,-0.044642
4,0.005383,-0.044642


In [ ]:
df[['age','sex']][-5:]

,age,sex
437,0.041708,0.050680
438,-0.005515,0.050680
439,0.041708,0.050680
440,-0.045472,-0.044642
441,-0.045472,-0.044642


* **USE df.loc[] and df.iloc to select only rows, only columns or both**



In [ ]:
df.iloc[0] # INteger base location
# it gives first row in this dataframe

,0
age,0.038076
sex,0.050680
bmi,0.061696
bp,0.021872
s1,-0.044223
s2,-0.034821
s3,-0.043401
s4,-0.002592
s5,0.019907
s6,-0.017646


In [ ]:
df.head(n=4)

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
0,0.038076,0.050680,0.061696,0.021872,-0.044223,-0.034821,-0.043401,-0.002592,0.019907,-0.017646
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204
2,0.085299,0.050680,0.044451,-0.005670,-0.045599,-0.034194,-0.032356,-0.002592,0.002861,-0.025930
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362


In [ ]:
df.loc[0 ,'age']  # Label-based selection
# here i can use

np.float64(0.038075906433423026)

in the loc specified lable of the row

In [ ]:
df.loc[0,'sex']

np.float64(0.05068011873981862)

* iloc position of the data in the row note here only uses integer it is similar to access from array
and in the loc is similar to access to like dictionary  

In [ ]:
df.iloc[0 , 2:7]
# first enteries is equal to row
# second enteries is equal to accessing column
# here i want zeroth row 2 to 6 number column data
# 2 = bmi , 3 = bp , 4 = s1, 5=s2, 6=s3

,0
bmi,0.061696
bp,0.021872
s1,-0.044223
s2,-0.034821
s3,-0.043401


In [ ]:
df.loc[4:6,'age']

,age
4,0.005383
5,-0.092695
6,-0.045472


* Label-based selection just like ['age', 'sex']

In [ ]:
df.loc[3,['age','sex']]

,3
age,-0.089063
sex,-0.044642


In [ ]:
age_df_temp = df.loc[df.age<5.383060e-03]

In [ ]:
age_df_temp

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362
5,-0.092695,-0.044642,-0.040696,-0.019442,-0.068991,-0.079288,0.041277,-0.076395,-0.041176,-0.096346
6,-0.045472,0.050680,-0.047163,-0.015999,-0.040096,-0.024800,0.000779,-0.039493,-0.062917,-0.038357
9,-0.070900,-0.044642,0.039062,-0.033213,-0.012577,-0.034508,-0.024993,-0.002592,0.067737,-0.013504
...,...,...,...,...,...,...,...,...,...,...
435,-0.012780,-0.044642,-0.023451,-0.040099,-0.016704,0.004636,-0.017629,-0.002592,-0.038460,-0.038357
436,-0.056370,-0.044642,-0.074108,-0.050427,-0.024960,-0.047034,0.092820,-0.076395,-0.061176,-0.046641
438,-0.005515,0.050680,-0.015906,-0.067642,0.049341,0.079165,-0.028674,0.034309,-0.018114,0.044485
440,-0.045472,-0.044642,0.039062,0.001215,0.016318,0.015283,-0.028674,0.026560,0.044529,-0.025930


In [ ]:
age_df_temp.iloc[0]


,1
age,-0.001882
sex,-0.044642
bmi,-0.051474
bp,-0.026328
s1,-0.008449
s2,-0.019163
s3,0.074412
s4,-0.039493
s5,-0.068332
s6,-0.092204


* below the code is give key error because in loc we put lable data which is in age_df_temp
it work like dict key but iloc is not give error if i put 2 index because it access as integer

In [ ]:
age_df_temp.loc[2]
# 2 lable is not present in Age_df_temp

KeyError: 2

* that why i called loc Label-based selection

In [ ]:
age_df_temp.iloc[2]

,5
age,-0.092695
sex,-0.044642
bmi,-0.040696
bp,-0.019442
s1,-0.068991
s2,-0.079288
s3,0.041277
s4,-0.076395
s5,-0.041176
s6,-0.096346


**Lecture-4 Data selection using pandas | part 2**

In [ ]:
import numpy as np

In [ ]:
df = pd.DataFrame(np.random.rand(9,4), index=list("abcdefghi"),columns=list('ABCD'))

In [ ]:
df.shape

(9, 4)

In [ ]:
df.columns

Index(['A', 'B', 'C', 'D'], dtype='object')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9 entries, a to i
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   A       9 non-null      float64
 1   B       9 non-null      float64
 2   C       9 non-null      float64
 3   D       9 non-null      float64
dtypes: float64(4)
memory usage: 360.0+ bytes


In [ ]:
df.head()

,A,B,C,D
a,0.360653,0.041973,0.950930,0.963117
b,0.470911,0.214636,0.521855,0.521809
c,0.103479,0.316955,0.109308,0.651124
d,0.069488,0.487899,0.594394,0.597303
e,0.368913,0.539537,0.249289,0.079812


In [ ]:
another_df = pd.DataFrame(
    np.random.rand(100,4),
    index=range(10,110),
    columns=list('abcd')
)

In [ ]:
another_df.head()

,a,b,c,d
10,0.249098,0.241393,0.156109,0.369985
11,0.288186,0.224101,0.846411,0.361528
12,0.058025,0.694821,0.376161,0.417380
13,0.391456,0.289799,0.195979,0.509747
14,0.039658,0.352239,0.442667,0.444608


In [ ]:
another_df.tail()

,a,b,c,d
105,0.260897,0.009894,0.660802,0.906584
106,0.746900,0.004589,0.327777,0.795066
107,0.333954,0.541284,0.749771,0.313446
108,0.866801,0.667104,0.043009,0.149693
109,0.922059,0.630394,0.942938,0.511140


In [ ]:
df.loc[0]
# it throw error because zero is not present in index

KeyError: 0

In [ ]:
df.index

Index(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i'], dtype='object')

In [ ]:
#df.loc[0]
# because in df 0 is not present but a is present

In [ ]:
df.loc['a']

,a
A,0.360653
B,0.041973
C,0.950930
D,0.963117


In [ ]:
df.loc['a','A':'C']


,a
A,0.360653
B,0.041973
C,0.950930


In [ ]:
df.loc['a', 'A':]

,a
A,0.360653
B,0.041973
C,0.950930
D,0.963117


In [ ]:
df.iloc[0:4,:] # row is 0:4 and column is : (all)

,A,B,C,D
a,0.360653,0.041973,0.950930,0.963117
b,0.470911,0.214636,0.521855,0.521809
c,0.103479,0.316955,0.109308,0.651124
d,0.069488,0.487899,0.594394,0.597303


In [ ]:
df.iloc[:,1]

lambda function which is use directly do not need to define function

In [ ]:
selector = lambda df: df['A']>0

In [ ]:
selector

<function __main__.<lambda>(df)>

In [ ]:
df.loc[selector]

,A,B,C,D
a,0.360653,0.041973,0.950930,0.963117
b,0.470911,0.214636,0.521855,0.521809
c,0.103479,0.316955,0.109308,0.651124
d,0.069488,0.487899,0.594394,0.597303
e,0.368913,0.539537,0.249289,0.079812
f,0.675259,0.827398,0.233085,0.441492
g,0.630261,0.477269,0.247173,0.113743
h,0.749426,0.895101,0.769015,0.493753
i,0.048654,0.318438,0.114589,0.129674


In [ ]:
selector = lambda df: df['A']>0.6

In [ ]:
print(selector)

<function <lambda> at 0x78ef3616a700>


In [ ]:
df.loc[selector]

,A,B,C,D
f,0.675259,0.827398,0.233085,0.441492
g,0.630261,0.477269,0.247173,0.113743
h,0.749426,0.895101,0.769015,0.493753


In [ ]:
selctor = lambda df: (df['A'] > 0.5) & (df['B']>0.2)

In [ ]:
selctor

<function __main__.<lambda>(df)>

In [ ]:
df.loc[selctor]

,A,B,C,D
f,0.675259,0.827398,0.233085,0.441492
g,0.630261,0.477269,0.247173,0.113743
h,0.749426,0.895101,0.769015,0.493753


In [ ]:
condition_for_selector = (df['A']>0.5) & (df['B']>0.4)

In [ ]:
df[condition_for_selector]

,A,B,C,D
f,0.675259,0.827398,0.233085,0.441492
g,0.630261,0.477269,0.247173,0.113743
h,0.749426,0.895101,0.769015,0.493753


In [ ]:
condition_for_selector

,0
a,False
b,False
c,False
d,False
e,False
f,True
g,True
h,True
i,False


**WEEK-1 LECTURE_5 :**
* **Drop in dataframe::**

**Adding a column in  the Dateframe**

Creating a new column E in my DataFrame df

np.random.rand.(9)
*   List item This comes from NumPy
*   List item 👉 It generates:

9 random numbers

Between 0 and 1



In [ ]:
df['E'] = np.random.rand(9)
# it creat 9 row within value  0 to 1

In [ ]:
df.head()

,A,B,C,D,E
a,0.360653,0.041973,0.950930,0.963117,0.056854
b,0.470911,0.214636,0.521855,0.521809,0.456654
c,0.103479,0.316955,0.109308,0.651124,0.568768
d,0.069488,0.487899,0.594394,0.597303,0.298338
e,0.368913,0.539537,0.249289,0.079812,0.916069


In [ ]:
df['F'] = df['E'] + df['A']

In [ ]:
df.head()

,A,B,C,D,E,F,M,G
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,67.391378,0.464017
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,63.367743,4.739908
c,0.103479,0.316955,0.109308,0.651124,0.568768,0.672247,55.381393,5.672214
d,0.069488,0.487899,0.594394,0.597303,0.298338,0.367826,74.552544,3.745112
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,33.231071,3.026398


In [ ]:
df['M'] = np.random.rand(9)*100

In [ ]:
df.head()

,A,B,C,D,E,F,M,G
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.464017
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,4.739908
c,0.103479,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,5.672214
d,0.069488,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,3.745112
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,3.026398


In [ ]:
df['G'] = df['E'] + df['F']

In [ ]:
df

,A,B,C,D,E,F,M,G
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219
c,0.103479,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015
d,0.069488,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225
h,0.749426,0.895101,0.769015,0.493753,0.970109,1.719535,92.372419,2.689644
i,0.048654,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403


In [ ]:
criteria = df['A']<0.2

In [ ]:
criteria

,A
a,False
b,False
c,True
d,True
e,False
f,False
g,False
h,False
i,True


In [ ]:
print(criteria.sum())

3


here you can see only one value is true

In [ ]:
df.loc[criteria]

,A,B,C,D,E,F,M,G
c,0.103479,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015
d,0.069488,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164
i,0.048654,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403


In [ ]:
df.loc[criteria,"A"]

,A
c,0.103479
d,0.069488
i,0.048654


In [ ]:
df.loc[criteria,"A"] = 0
# noone data is equal to 0 so out put is nothing

In [ ]:
df

,A,B,C,D,E,F,M,G
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225
h,0.749426,0.895101,0.769015,0.493753,0.970109,1.719535,92.372419,2.689644
i,0.000000,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403


In [ ]:
cities = ['Mumbai','Delhi','Chennai','Kolkata','Bengalore','Hydrabad','Pune','Ahmdabad','Indore']

In [ ]:
df['City'] = cities

In [ ]:
df

,A,B,C,D,E,F,M,G,city,City
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360,Mumbai,Mumbai
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219,Delhi,Delhi
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015,Chennai,Chennai
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164,Kolkata,Kolkata
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,Bengalore,Bengalore
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967,Hydrabad,Hydrabad
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225,Pune,Pune
h,0.749426,0.895101,0.769015,0.493753,0.970109,1.719535,92.372419,2.689644,Ahmdabad,Ahmdabad
i,0.000000,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403,Indore,Indore


In [ ]:
cities_new = ['Mumbai','Delhi','Chennai','Kolkata','Bengalore','Hydirabad','Pune','Ahmdabad','NA']

In [ ]:
df['New_City'] = cities_new

In [ ]:
df

,A,B,C,D,E,F,M,G,city,City,New_City
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360,Mumbai,Mumbai,Mumbai
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219,Delhi,Delhi,Delhi
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015,Chennai,Chennai,Chennai
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164,Kolkata,Kolkata,Kolkata
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,Bengalore,Bengalore,Bengalore
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967,Hydrabad,Hydrabad,Hydirabad
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225,Pune,Pune,Pune
h,0.749426,0.895101,0.769015,0.493753,0.970109,1.719535,92.372419,2.689644,Ahmdabad,Ahmdabad,Ahmdabad
i,0.000000,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403,Indore,Indore,NA


In [ ]:
df_copy = df.copy()

In [ ]:
df_copy['new_city'] = cities_new

In [ ]:
df_copy

,A,B,C,D,E,F,M,G,city,City,New_City,new_city
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360,Mumbai,Mumbai,Mumbai,Mumbai
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219,Delhi,Delhi,Delhi,Delhi
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015,Chennai,Chennai,Chennai,Chennai
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164,Kolkata,Kolkata,Kolkata,Kolkata
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,Bengalore,Bengalore,Bengalore,Bengalore
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967,Hydrabad,Hydrabad,Hydirabad,Hydirabad
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225,Pune,Pune,Pune,Pune
h,0.749426,0.895101,0.769015,0.493753,0.970109,1.719535,92.372419,2.689644,Ahmdabad,Ahmdabad,Ahmdabad,Ahmdabad
i,0.000000,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403,Indore,Indore,NA,NA


In [ ]:
criteria = df['city'].isin(['Pune','Bengalore','Hyderabad'])

In [ ]:
criteria

,city
a,False
b,False
c,False
d,False
e,True
f,False
g,True
h,False
i,False


** how i can change one perticular inteties in this

In [ ]:
df_copy.loc[df.city == "Bengalore", ['city','new_cities']] = 'bengalore'

In [ ]:
df_copy

,A,B,C,D,E,F,M,G,city,City,New_City,new_city,new_cities
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360,Mumbai,Mumbai,Mumbai,Mumbai,NaN
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219,Delhi,Delhi,Delhi,Delhi,NaN
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015,Chennai,Chennai,Chennai,Chennai,NaN
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164,Kolkata,Kolkata,Kolkata,Kolkata,NaN
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,bengalore,Bengalore,Bengalore,Bengalore,bengalore
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967,Hydrabad,Hydrabad,Hydirabad,Hydirabad,NaN
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225,Pune,Pune,Pune,Pune,NaN
h,0.749426,0.895101,0.769015,0.493753,0.970109,1.719535,92.372419,2.689644,Ahmdabad,Ahmdabad,Ahmdabad,Ahmdabad,NaN
i,0.000000,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403,Indore,Indore,NA,NA,NaN


* **how to drop a column and row

In [ ]:
?df_copy.drop

In [ ]:
df_copy.drop(['new_cities'],axis=1)
# axis = 1 means do operation on the column and if axis = 0 it means operation on the row

,A,B,C,D,E,F,M,G,city,City,New_City,new_city
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360,Mumbai,Mumbai,Mumbai,Mumbai
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219,Delhi,Delhi,Delhi,Delhi
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015,Chennai,Chennai,Chennai,Chennai
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164,Kolkata,Kolkata,Kolkata,Kolkata
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,bengalore,Bengalore,Bengalore,Bengalore
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967,Hydrabad,Hydrabad,Hydirabad,Hydirabad
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225,Pune,Pune,Pune,Pune
h,0.749426,0.895101,0.769015,0.493753,0.970109,1.719535,92.372419,2.689644,Ahmdabad,Ahmdabad,Ahmdabad,Ahmdabad
i,0.000000,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403,Indore,Indore,NA,NA


In [ ]:
df_copy.loc[:,'city'] = 'Chainnai'

In [ ]:
df_copy

,A,B,C,D,E,F,M,G,city,City,New_City,new_city,new_cities
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360,Chainnai,Mumbai,Mumbai,Mumbai,NaN
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219,Chainnai,Delhi,Delhi,Delhi,NaN
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015,Chainnai,Chennai,Chennai,Chennai,NaN
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164,Chainnai,Kolkata,Kolkata,Kolkata,NaN
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,Chainnai,Bengalore,Bengalore,Bengalore,bengalore
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967,Chainnai,Hydrabad,Hydirabad,Hydirabad,NaN
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225,Chainnai,Pune,Pune,Pune,NaN
h,0.749426,0.895101,0.769015,0.493753,0.970109,1.719535,92.372419,2.689644,Chainnai,Ahmdabad,Ahmdabad,Ahmdabad,NaN
i,0.000000,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403,Chainnai,Indore,NA,NA,NaN


** how drop row

In [ ]:
df_copy.drop(['i'])

,A,B,C,D,E,F,M,G,city,City,New_City,new_city,new_cities
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360,Chainnai,Mumbai,Mumbai,Mumbai,NaN
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219,Chainnai,Delhi,Delhi,Delhi,NaN
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015,Chainnai,Chennai,Chennai,Chennai,NaN
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164,Chainnai,Kolkata,Kolkata,Kolkata,NaN
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,Chainnai,Bengalore,Bengalore,Bengalore,bengalore
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967,Chainnai,Hydrabad,Hydirabad,Hydirabad,NaN
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225,Chainnai,Pune,Pune,Pune,NaN
h,0.749426,0.895101,0.769015,0.493753,0.970109,1.719535,92.372419,2.689644,Chainnai,Ahmdabad,Ahmdabad,Ahmdabad,NaN


In [ ]:
?df_copy.sample

In [ ]:
df.sample(3)
# here it choose randomly 3 sample

,A,B,C,D,E,F,M,G,city,City,New_City
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164,Kolkata,Kolkata,Kolkata
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360,Mumbai,Mumbai,Mumbai
b,0.470911,0.214636,0.521855,0.521809,0.456654,0.927565,46.916387,1.384219,Delhi,Delhi,Delhi


In [ ]:
df_copy.sample(3)
# here it choose randomly 3 sample

,A,B,C,D,E,F,M,G,city,City,New_City,new_city,new_cities
i,0.000000,0.318438,0.114589,0.129674,0.424375,0.473029,39.100809,0.897403,Chainnai,Indore,NA,NA,NaN
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225,Chainnai,Pune,Pune,Pune,NaN
a,0.360653,0.041973,0.950930,0.963117,0.056854,0.417507,89.767697,0.474360,Chainnai,Mumbai,Mumbai,Mumbai,NaN


random
-state it stop sample generation radomly after this it generate same sample

In [ ]:
df.sample(3,random_state=78)
# here it choose randomly 3 sample but i fix it by using random_satate

,A,B,C,D,E,F,M,G,city,City,New_City
d,0.000000,0.487899,0.594394,0.597303,0.298338,0.367826,93.283431,0.666164,Kolkata,Kolkata,Kolkata
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015,Chennai,Chennai,Chennai
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,Bengalore,Bengalore,Bengalore


In [ ]:
df.sample(3,random_state=23)

,A,B,C,D,E,F,M,G,city,City,New_City
c,0.000000,0.316955,0.109308,0.651124,0.568768,0.672247,11.699684,1.241015,Chennai,Chennai,Chennai
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,Bengalore,Bengalore,Bengalore
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967,Hydrabad,Hydrabad,Hydirabad


In [ ]:
df_copy.sample(3, replace = True)
# repetation is allowed

,A,B,C,D,E,F,M,G,city,City,New_City,new_city,new_cities
f,0.675259,0.827398,0.233085,0.441492,0.511354,1.186613,56.210590,1.697967,Chainnai,Hydrabad,Hydirabad,Hydirabad,NaN
g,0.630261,0.477269,0.247173,0.113743,0.728482,1.358743,67.569116,2.087225,Chainnai,Pune,Pune,Pune,NaN
e,0.368913,0.539537,0.249289,0.079812,0.916069,1.284982,43.266442,2.201051,Chainnai,Bengalore,Bengalore,Bengalore,bengalore


**DataFrame.sample(n=None, frac=None, replace=False, weights=None, random_state=None, axis=None, ignore_index=False)**
these are following attribute of sample function
1. n👉 Number of rows to select
2. frac👉 Fraction (percentage) of data
3. replace👉 Allow repetition or not
False → no repeat (default)
True → repeat allowed
4. weights👉 Probability for each row
5. random_state👉 Fix randomness (for same result every time)
6. axis👉 Sample rows or columns
0 → rows (default)
1 → columns
ignore_index
👉 Reset index or not

* **WEEK-1 Lecture-6:**
* **Apply statistics using pandas**

In [ ]:
import pandas as pd
cities = ['Mumbai','Chennai','Pune','Ahmedabad','Kolkata','Kanpur', 'Delhi']
city_df = pd.DataFrame(cities)

In [ ]:
city_df.columns = ['city_name']

In [ ]:
city_df

,city_name
0,Mumbai
1,Chennai
2,Pune
3,Ahmedabad
4,Kolkata
5,Kanpur
6,Delhi


In [ ]:
condition = city_df.city_name == 'Mumbai'

In [ ]:
condition

,city_name
0,True
1,False
2,False
3,False
4,False
5,False
6,False


In [ ]:
type(condition)


pandas.core.series.Series

In [ ]:
city_df[condition]

,city_name
0,Mumbai


In [ ]:
city_df[city_df.city_name == 'Pune']

,city_name
2,Pune


* **Aggregation and grouping**

In [ ]:
import numpy as np

In [ ]:
random_state = np.random.RandomState(43)
random_series = pd.Series(random_state.rand(10))

In [ ]:
random_series

,0
0,0.115055
1,0.609067
2,0.133391
3,0.240590
4,0.327139
5,0.859137
6,0.666090
7,0.541162
8,0.029014
9,0.733748


* **Now that the series is created,**
* **we can try aggregation function like mean,
std etc.**

In [ ]:
random_series.mean()

np.float64(0.42543927826664074)

In [ ]:
random_series.std()

0.2926485546102095

In [ ]:
random_series.sum()

np.float64(4.254392782666407)

* **let's created a dataframe with 5 rows and 2 columns contaning random number**

In [ ]:
df = pd.DataFrame(random_state.rand(5,2), columns = ['A','B'])
# this a secon way ({"A":random_state.rand(5), "B":random_state.rand(5)})

In [ ]:
df

,A,B
0,0.848939,0.971465
1,0.385377,0.954488
2,0.445758,0.669725
3,0.082500,0.897099
4,0.298004,0.262305


In [ ]:
df.sum()

,0
A,2.060578
B,3.755081


In [ ]:
df.mean()


,0
A,0.412116
B,0.751016


**Try these operation row_wise**

In [ ]:
df.mean(axis=1)

,0
0,0.910202
1,0.669933
2,0.557742
3,0.489799
4,0.280154


In [ ]:
df.sum(axis=1)

,0
0,1.820404
1,1.339865
2,1.115483
3,0.979599
4,0.560308


## Groupby
* ## group have three operation
* **Split** - we split datafram into smaller dataframe based on the value of keys
* **Apply** - we apply desired
aggregation/trandsformation on each dataframe
* **Combine**- we combine result from apply state into a datafram



In [ ]:
df = pd.DataFrame({'key':['A','B','C']*2, # string ABCABC
                   'data': range(6)})

In [ ]:
df

,key,data
0,A,0
1,B,1
2,C,2
3,A,3
4,B,4
5,C,5


In [ ]:
df.groupby('key').sum()

,data
key,
A,3
B,5
C,7


In [ ]:
df.groupby(by='key').sum()

,data
key,
A,3
B,5
C,7


In [ ]:
df.groupby(level='data').sum()

ValueError: level name data is not the name of the index